### PR TITLE
feat: added global loading template token

### PR DESCRIPTION
**Pull request by itayod**

### CODE REVIEW

File: projects/transloco/src/lib/loaderComponent.component.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { Component, Input, ChangeDetectorRef } from '@angular/core';
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/loaderComponent.component.ts]: file names - dash case

File: projects/transloco/src/lib/loaderComponent.component.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { Component, Input, ChangeDetectorRef } from '@angular/core';
+
+@Component({
+  template: `
+    <div class="transloco-loader-template" [innerHTML]="html"></div>
+  `
+})
+export class LoaderComponent {
+  private _html: string;
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/loaderComponent.component.ts]: We don't need getter and setter. It should be:
`@Input() html: string;`

File: projects/transloco/src/lib/loaderComponent.component.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { Component, Input, ChangeDetectorRef } from '@angular/core';
+
+@Component({
+  template: `
+    <div class="transloco-loader-template" [innerHTML]="html"></div>
+  `
+})
+export class LoaderComponent {
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/loaderComponent.component.ts]: prefix with Transloco

File: projects/transloco/src/lib/templateHandler.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import {
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: dash case

File: projects/transloco/src/lib/templateHandler.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  ViewContainerRef,
+  ComponentFactoryResolver,
+  Component,
+  Injector
+} from '@angular/core';
+import { LoaderComponent } from './loaderComponent.component';
+
+export class TemplateHandler {
+  constructor(
+    private template: string | TemplateRef<any> | Component,
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: `Component => Type<any>`

File: projects/transloco/src/lib/templateHandler.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  private createComponent<T>(cmp): ComponentRef<T> {
+    const cfr = this.injector.get(ComponentFactoryResolver);
+    const component = cfr.resolveComponentFactory(cmp);
+
+    return this.vcr.createComponent(component);
+  }
+
+  private isTemplateRef(template: string | TemplateRef<any> | Component) {
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: DRY. extract `string | TemplateRef<any> | Component` to a type

File: projects/transloco/src/lib/templateHandler.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    return template instanceof TemplateRef;
+  }
+
+  private isHTML(template: string | TemplateRef<any> | Component) {
+    return typeof template === 'string';
+  }
+
+  private isComponent(template: string | TemplateRef<any> | Component) {
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: Redundant, the other checks are enough.

File: projects/transloco/src/lib/templateHandler.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    return this.vcr.createComponent(component);
+  }
+
+  private isTemplateRef(template: string | TemplateRef<any> | Component) {
+    return template instanceof TemplateRef;
+  }
+
+  private isHTML(template: string | TemplateRef<any> | Component) {
+    return typeof template === 'string';
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: we have `isString` in our helpers

File: projects/transloco/src/lib/templateHandler.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      const loaderCmp = this.createComponent<LoaderComponent>(LoaderComponent);
+      loaderCmp.instance.html = this.template as string;
+    }
+  }
+
+  public detachView() {
+    this.vcr.clear();
+  }
+
+  private createComponent<T>(cmp): ComponentRef<T> {
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: `cmp: Type<T>`

File: projects/transloco/src/lib/templateHandler.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    const cfr = this.injector.get(ComponentFactoryResolver);
+    const component = cfr.resolveComponentFactory(cmp);
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: component => factory

File: projects/transloco/src/lib/templateHandler.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  ) {}
+
+  public attachView() {
+    if (this.isTemplateRef(this.template)) {
+      this.vcr.createEmbeddedView(this.template as TemplateRef<any>);
+    } else if (this.isComponent(this.template)) {
+      this.createComponent(this.template);
+    } else if (this.isHTML(this.template)) {
+      const loaderCmp = this.createComponent<LoaderComponent>(LoaderComponent);
+      loaderCmp.instance.html = this.template as string;
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: run `detectChanges()`

File: projects/transloco/src/lib/transloco.directive.ts [hunk 12]
```diff
@@ -82,16 +94,20 @@ export class TranslocoDirective implements OnInit, OnDestroy, OnChanges {
...
-      this.hasLoadingTpl() && this.vcr.clear();
+      this.loaderTplHandler.detachView();
       this.view = this.vcr.createEmbeddedView(this.tpl, {
         $implicit: data
       });
     }
   }
 
+  private getLoadingTpl(): string | TemplateRef<any> | Component {
+    return this.loadingTpl || this.defaultLoading;
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/transloco.directive.ts]: Be consistent with the names. 

loadingTpl => inlineLoadingTpl
defaultLoading => providerLoadingTpl

File: projects/transloco/src/lib/transloco.directive.ts [hunk 13]
```diff
@@ -33,22 +39,28 @@ export class TranslocoDirective implements OnInit, OnDestroy, OnChanges {
...
     private vcr: ViewContainerRef,
     private cdr: ChangeDetectorRef,
     private host: ElementRef
-  ) {
-  }
+  ) {}
 
   ngOnInit() {
-    this.hasLoadingTpl() && this.vcr.createEmbeddedView(this.loadingTpl);
+    const tpl = this.getLoadingTpl();
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/transloco.directive.ts]: tpl => loadingTpl

File: projects/transloco/src/lib/loaderComponent.component.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import { Component, Input, ChangeDetectorRef } from '@angular/core';
+
+@Component({
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/loaderComponent.component.ts]: You need to add it to entry components

File: projects/transloco/src/lib/transloco.directive.ts [hunk 15]
```diff
@@ -33,22 +39,28 @@ export class TranslocoDirective implements OnInit, OnDestroy, OnChanges {
...
+    if (tpl) {
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/transloco.directive.ts]: What about when we load new lang? Do we need to show it while it loads?

File: projects/transloco/src/lib/templateHandler.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+  Component,
+  Injector
+} from '@angular/core';
+import { LoaderComponent } from './loaderComponent.component';
+
+export class TemplateHandler {
+  constructor(
+    private template: string | TemplateRef<any> | Component,
+    private vcr: ViewContainerRef,
+    private injector: Injector
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/templateHandler.ts]: Why you pass the injector? pass directly the cfr.

File: src/app/lazy/lazy.component.html [hunk 17]
```diff
@@ -1,4 +1,4 @@
-<ng-container *transloco="let t; loadingTpl: loading">
+<ng-container *transloco="let t">
```
Comment (NetanelBasal) [inline src/app/lazy/lazy.component.html]: add also an inline example, with the text: inline loader wins:

**Comment by NetanelBasal** [comment]
The ci is broken.

File: projects/transloco/src/lib/loaderComponent.component.ts (hunk 14 above)
Comment (itayod) [reply projects/transloco/src/lib/loaderComponent.component.ts]: It's on entry components already

File: projects/transloco/src/lib/templateHandler.ts (hunk 7 above)
Comment (itayod) [reply projects/transloco/src/lib/templateHandler.ts]: so how will I know how to refer it?

Comment (NetanelBasal) [reply projects/transloco/src/lib/templateHandler.ts]: The last else is a component

File: projects/transloco/src/lib/templateHandler.ts (hunk 16 above)
Comment (itayod) [reply projects/transloco/src/lib/templateHandler.ts]: I don't want to pass to many dependencies and force every consumer to inject those dependencies to its class.

File: src/app/lazy/lazy.component.html (hunk 17 above)
Comment (itayod) [reply src/app/lazy/lazy.component.html]: there is one, just below...

File: projects/transloco/src/lib/transloco.directive.ts (hunk 13 above)
Comment (NetanelBasal) [reply projects/transloco/src/lib/transloco.directive.ts]: ?

File: projects/transloco/src/lib/templateHandler.ts (hunk 7 above)
Comment (itayod) [reply projects/transloco/src/lib/templateHandler.ts]: ok

File: projects/transloco/src/lib/template-handler.ts [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+  public attachView() {
+    if (this.isTemplateRef(this.template)) {
+      this.vcr.createEmbeddedView(this.template as TemplateRef<any>);
+    } else if (this.isComponent(this.template)) {
+      this.createComponent(this.template as Type<any>);
+    } else if (this.isHTML(this.template)) {
+      const loaderCmp = this.createComponent<TranslocoLoaderComponent>(TranslocoLoaderComponent);
+      loaderCmp.instance.html = this.template as string;
+      const cdr = this.injector.get(ChangeDetectorRef);
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/template-handler.ts]: You don't need to grab the injector. 

`
const componentRef = this.createComponent<TranslocoLoaderComponent>(TranslocoLoaderComponent);
componentRef.hostView.detectChanges();
`

File: projects/transloco/src/lib/template-handler.ts [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+  ComponentFactoryResolver,
+  Component,
+  Injector,
+  Type,
+  ChangeDetectorRef
+} from '@angular/core';
+import { isString } from './helpers';
+import { TranslocoLoaderComponent } from './loader-component.component';
+
+export type Template = string | TemplateRef<any> | Type<any>;
```
Comment (NetanelBasal) [inline projects/transloco/src/lib/template-handler.ts]: `View` makes more sense.

**Pull request merged**